src/test/unit/io/deserializer_test.cpp

```diff
@@ -0,0 +1,899 @@
+  EXPECT_FLOAT_EQ(5.0, y[2](1));
+  EXPECT_FLOAT_EQ(6.0, y[3](0));
+  EXPECT_FLOAT_EQ(7.0, y[3](1));
+
+  double z = deserializer.read<double>();
+  EXPECT_FLOAT_EQ(8.0, z);
+}
+
+// size zero
+
+TEST(deserializer, zeroSizeVecs) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(1.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  EXPECT_FLOAT_EQ(1.0, deserializer.read<double>());  // finish available
+
+  EXPECT_EQ(0, deserializer.read<std::vector<double>>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::VectorXd>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::RowVectorXd>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::MatrixXd>(0, 3).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::MatrixXd>(3, 0).size());
+  EXPECT_EQ(0, deserializer
+                   .read<std::vector<std::vector<Eigen::MatrixXd>>>(0, 0, 0, 0)
+                   .size());
+}
+
+// out of memory
+
+TEST(deserializer, eos_exception) {
+  std::vector<double> theta;
+  theta.push_back(1.0);
+  theta.push_back(2.0);
+  std::vector<int> theta_i;
+  theta_i.push_back(1);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  EXPECT_EQ(2U, deserializer.available());
+  EXPECT_EQ(1U, deserializer.available_i());
+
+  EXPECT_NO_THROW(deserializer.read<double>());
+  EXPECT_NO_THROW(deserializer.read<double>());
+  EXPECT_THROW(deserializer.read<double>(), std::runtime_error);
+
+  // should go back to working
+  EXPECT_NO_THROW(deserializer.read<int>());
+  EXPECT_THROW(deserializer.read<int>(), std::runtime_error);
+
+  // should keep throwing
+  EXPECT_THROW(deserializer.read<double>(), std::runtime_error);
+  EXPECT_THROW(deserializer.read<size_t>(), std::runtime_error);
+
+  EXPECT_THROW(deserializer.read<std::complex<double>>(), std::runtime_error);
+  EXPECT_THROW(deserializer.read<std::vector<double>>(3), std::runtime_error);
+  EXPECT_THROW(deserializer.read<Eigen::VectorXd>(2), std::runtime_error);
+  EXPECT_THROW(
+      (deserializer
+           .read<Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>>(2)),
+      std::runtime_error);
+  EXPECT_THROW(deserializer.read<std::vector<Eigen::VectorXd>>(3, 2),
+               std::runtime_error);
+  EXPECT_THROW(deserializer.read<Eigen::RowVectorXd>(2), std::runtime_error);
+  EXPECT_THROW(
+      (deserializer
+           .read<Eigen::Matrix<std::complex<double>, 1, Eigen::Dynamic>>(2)),
+      std::runtime_error);
+  EXPECT_THROW(deserializer.read<std::vector<Eigen::RowVectorXd>>(3, 2),
+               std::runtime_error);
+  EXPECT_THROW(deserializer.read<Eigen::MatrixXd>(2, 2), std::runtime_error);
+  EXPECT_THROW(
+      (deserializer.read<
+          Eigen::Matrix<std::complex<double>, Eigen::Dynamic, Eigen::Dynamic>>(
+          2, 2)),
+      std::runtime_error);
+  EXPECT_THROW(deserializer.read<std::vector<Eigen::MatrixXd>>(2, 3, 2),
+               std::runtime_error);
+}
+
+// lb
+
+TEST(deserializer_scalar, read_lb_constrain) {
```

Comment:
For lb/ub/lub/multiplier_offset I went pretty sparse with the tests. I didn't check that all combinations of vector/`std::vector` etc. etc. worked. I'm assuming these functions work in Math and we're just trying to test the deserializer interface to them.